vm/src/vm/vm_ops.rs

```diff
@@ -174,160 +177,346 @@ impl VirtualMachine {
+                let ret = seq_a.concat(b, vm)?;
+                if ret.rich_compare_bool(
+                    vm.ctx.not_implemented.as_object(),
+                    PyComparisonOp::Ne,
+                    vm,
+                )? {
+                    return Ok(ret);
+                }
+            }
+
+            Ok(vm.ctx.not_implemented())
+        })
+    }
+
+    pub fn _iadd(&self, a: &PyObject, b: &PyObject) -> PyResult {
+        self.binary_iop(
             a,
             b,
-            identifier!(self, __add__),
-            identifier!(self, __radd__),
-            |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "+")),
+            PyNumberMethodsOffset::InplaceAdd,
+            PyNumberMethodsOffset::Add,
+            "+=",
+            |vm, a, b| {
+                let seq_a = PySequence::new(a, vm);
+
+                if let Some(seq_a) = seq_a {
+                    return seq_a.inplace_concat(b, vm);
+                }
+
+                Ok(vm.ctx.not_implemented())
+            },
         )
     }
 
-    pub fn _iadd(&self, a: &PyObject, b: &PyObject) -> PyResult {
-        self.call_or_unsupported(a, b, identifier!(self, __iadd__), |vm, a, b| {
-            vm.call_or_reflection(
-                a,
-                b,
-                identifier!(self, __add__),
-                identifier!(self, __radd__),
-                |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "+=")),
-            )
+    pub fn _sub(&self, a: &PyObject, b: &PyObject) -> PyResult {
+        self.binary_op(a, b, PyNumberMethodsOffset::Subtract, "-", |vm, _, _| {
+            Ok(vm.ctx.not_implemented())
         })
     }
 
-    pub fn _mul(&self, a: &PyObject, b: &PyObject) -> PyResult {
-        self.call_or_reflection(
+    pub fn _isub(&self, a: &PyObject, b: &PyObject) -> PyResult {
+        self.binary_iop(
             a,
             b,
-            identifier!(self, __mul__),
-            identifier!(self, __rmul__),
-            |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "*")),
+            PyNumberMethodsOffset::InplaceSubtract,
+            PyNumberMethodsOffset::Subtract,
+            "-=",
+            |vm, _, _| Ok(vm.ctx.not_implemented()),
         )
     }
 
-    pub fn _imul(&self, a: &PyObject, b: &PyObject) -> PyResult {
-        self.call_or_unsupported(a, b, identifier!(self, __imul__), |vm, a, b| {
-            vm.call_or_reflection(
-                a,
-                b,
-                identifier!(self, __mul__),
-                identifier!(self, __rmul__),
-                |vm, a, b| Err(vm.new_unsupported_binop_error(a, b, "*=")),
-            )
+    pub fn _mul(&self, a: &PyObject, b: &PyObject) -> PyResult {
+        self.binary_op(a, b, PyNumberMethodsOffset::Multiply, "*", |vm, a, b| {
+            let seq_a = PySequence::new(a, vm);
+            let seq_b = PySequence::new(b, vm);
+
+            // TODO: I think converting to isize process should be handled in repeat function.
```

Comment:
who is "I"? the future reader will be confused.

If you think this topic requires discussion rather than leaving concrete TODO lists,
please consider opening an issue and leave here the issue number or link.

Reply:
If allowing arbitary n is not necessary for cpython compatibility, I think `repeat` is better to take `isize` or `usize` because we can skip range check when we already know repeating `n` is possible.
If we are going to remove repetitive pattern, adding `try_repeat` with more inclusive `n` will be possible.

Reply:
another idea is this will be a temporary code.
once all sequences implement number protocol correctly, they will handle it in their own number protocol.